### PR TITLE
feat: Add data persistence via DBFlow

### PR DESCRIPTION
**Pull request by iamareebjamal**
Fixes part of #156

- Add DbFlow
- Add stetho
- Implement unit tested persistence API for loading evemt

### CODE REVIEW

**Comment by iamareebjamal** [comment]
@arp95 @Shailesh351 @mejariamol @mahikaw @rohanagarwal94 Please review

**Comment by iamareebjamal** [comment]
Testing Java is so hard, the whole day was spent just rewriting tests when the functionality was complete yesterday only

**Comment by iamareebjamal** [comment]
@mejariamol Please review this as I cannot move forward unless this gets merged

**Comment by mejariamol** [comment]
Yes. I am looking into the code already. Will add my review by tonight. :)

File: app/src/main/java/org/fossasia/openevent/app/OrgaApplication.java [hunk 1]
```diff
@@ -66,11 +83,13 @@ public void onCreate() {
 
         appComponent = DaggerAppComponent.builder()
             .androidModule(new AndroidModule(this))
-            .dataModule(new DataModule())
-            .networkModule(new NetworkModule())
             .build();
     }
 
+    public boolean isTestBuild() {
+        return false;
```
Comment (mejariamol) [inline app/src/main/java/org/fossasia/openevent/app/OrgaApplication.java]: What is purpose of it returning `false` explicitly?

Comment (mejariamol) [inline app/src/main/java/org/fossasia/openevent/app/OrgaApplication.java]: What is purpose of it returning `false` explicitly?

File: app/src/main/java/org/fossasia/openevent/app/common/di/module/DataModule.java [hunk 2]
```diff
@@ -23,7 +24,8 @@ ICacheModel providesCacheModel() {
 
     @Provides
     @Singleton
-    IEventRepository providesEventRepository(IUtilModel utilModel, ICacheModel cacheModel, EventService eventService) {
-        return new EventRepository(utilModel, cacheModel, eventService);
+    IEventRepository providesEventRepository(IUtilModel utilModel, IDatabaseRepository databaseRepository,
+                                             ICacheModel cacheModel, EventService eventService) {
```
Comment (mejariamol) [inline app/src/main/java/org/fossasia/openevent/app/common/di/module/DataModule.java]: why do we need cache model as now we have local database?

File: app/src/main/java/org/fossasia/openevent/app/data/EventRepository.java [hunk 3]
```diff
@@ -63,27 +69,74 @@ public EventRepository(IUtilModel utilModel, ICacheModel cacheModel, EventServic
...
+        Observable<Event> diskObservable = Observable.defer(() ->
+            databaseRepository.getAllItems(Event.class)
+        );
+
+        Observable<Event> networkObservable = Observable.defer(() ->
+            eventService.getEvents(authorization)
+                    .flatMapIterable(events -> events)
+                    .doOnNext(event -> {
+                        event.setComplete(false);
+                        databaseRepository.save(event).subscribe();
```
Comment (mejariamol) [inline app/src/main/java/org/fossasia/openevent/app/data/EventRepository.java]: how **updating** _already present data_ is managed here?

File: app/src/main/java/org/fossasia/openevent/app/data/db/DatabaseRepository.java [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+import com.raizlabs.android.dbflow.sql.language.SQLOperator;
+import com.raizlabs.android.dbflow.sql.language.SQLite;
+import com.raizlabs.android.dbflow.structure.BaseModel;
+
+import org.fossasia.openevent.app.data.db.contract.IDatabaseRepository;
+
+import io.reactivex.Completable;
+import io.reactivex.Observable;
+
+public class DatabaseRepository implements IDatabaseRepository {
```
Comment (mejariamol) [inline app/src/main/java/org/fossasia/openevent/app/data/db/DatabaseRepository.java]: How are **updates** handled? 
For example - when user requests reload, how its gonna update the stored database?

File: app/src/main/java/org/fossasia/openevent/app/OrgaApplication.java (hunk 1 above)
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/app/OrgaApplication.java]: Please check `TestOrgaApplication`

File: app/src/main/java/org/fossasia/openevent/app/common/di/module/DataModule.java (hunk 2 above)
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/app/common/di/module/DataModule.java]: How will the attendees load?

The database only stores event

File: app/src/main/java/org/fossasia/openevent/app/data/EventRepository.java (hunk 3 above)
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/app/data/EventRepository.java]: Please see corresponding tests for updating on reload. That's what tests are for. Besides `databaseRepository.save(event).subscribe()` should give a clue

File: app/src/main/java/org/fossasia/openevent/app/data/db/DatabaseRepository.java (hunk 4 above)
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/app/data/db/DatabaseRepository.java]: Please see corresponding tests for `shouldUpdateEventsOnReload` and `shouldUpdateEventOnReload` and `getAbstractObservable`

The basic idea of tests is if these are failing, the functionality doesn't work, but if they are passing, it definitely works

File: app/src/main/java/org/fossasia/openevent/app/data/db/configuration/DbFlowExclusionStrategy.java [hunk 5]
```diff
@@ -0,0 +1,18 @@
+package org.fossasia.openevent.app.data.db.configuration;
+
+import com.google.gson.ExclusionStrategy;
+import com.google.gson.FieldAttributes;
+import com.raizlabs.android.dbflow.structure.ModelAdapter;
+
+public class DbFlowExclusionStrategy implements ExclusionStrategy {
```
Comment (rohanagarwal94) [inline app/src/main/java/org/fossasia/openevent/app/data/db/configuration/DbFlowExclusionStrategy.java]: This class seems unused. Any specific reason to create it?

File: app/src/main/java/org/fossasia/openevent/app/data/models/Event.java [hunk 6]
```diff
@@ -380,26 +447,78 @@ public void setVersion(Version version) {
...
+    // One to Many implementation
+
+    @OneToMany(methods = {OneToMany.Method.ALL}, variableName = "tickets")
+    public List<Ticket> getEventTickets() {
+        if(tickets != null && !tickets.isEmpty())
+            return tickets;
+
+        tickets = SQLite.select()
+            .from(Ticket.class)
+            .where(Ticket_Table.event_id.eq(id))
```
Comment (rohanagarwal94) [inline app/src/main/java/org/fossasia/openevent/app/data/models/Event.java]: `Ticket_table` cannot be resolved in the studio.

File: app/src/main/java/org/fossasia/openevent/app/data/models/Event.java [hunk 7]
```diff
@@ -380,26 +447,78 @@ public void setVersion(Version version) {
...
+    }
+
+    @OneToMany(methods = {OneToMany.Method.ALL}, variableName = "socialLinks")
+    public List<SocialLink> getEventSocialLinks() {
+        if(socialLinks != null && !socialLinks.isEmpty())
+            return socialLinks;
+
+        socialLinks = SQLite.select()
+            .from(SocialLink.class)
+            .where(SocialLink_Table.event_id.eq(id))
```
Comment (rohanagarwal94) [inline app/src/main/java/org/fossasia/openevent/app/data/models/Event.java]: Same here with `SocialLink_Table`.

File: app/src/main/java/org/fossasia/openevent/app/data/models/SimpleModel.java [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+@Table(database = OrgaDatabase.class, allFields = true)
+public class SimpleModel extends BaseModel {
+
+    @PrimaryKey
+    public long id;
+
+    public String name;
+    public String description;
+
+    SimpleModel() {}
```
Comment (rohanagarwal94) [inline app/src/main/java/org/fossasia/openevent/app/data/models/SimpleModel.java]: Why not public?

File: app/src/main/java/org/fossasia/openevent/app/data/db/configuration/DbFlowExclusionStrategy.java (hunk 5 above)
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/app/data/db/configuration/DbFlowExclusionStrategy.java]: Thanks for reminding. This needs to be added in Gson Exclusion strategy

File: app/src/main/java/org/fossasia/openevent/app/data/EventRepository.java [hunk 9]
```diff
@@ -1,10 +1,14 @@
...
 
+import com.raizlabs.android.dbflow.structure.BaseModel;
+
 import org.fossasia.openevent.app.data.cache.ICacheModel;
 import org.fossasia.openevent.app.data.contract.IEventRepository;
 import org.fossasia.openevent.app.data.contract.IUtilModel;
+import org.fossasia.openevent.app.data.db.contract.IDatabaseRepository;
 import org.fossasia.openevent.app.data.models.Attendee;
 import org.fossasia.openevent.app.data.models.Event;
+import org.fossasia.openevent.app.data.models.Event_Table;
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/app/data/EventRepository.java]: It is autogenerated by DBFlow

File: app/src/main/java/org/fossasia/openevent/app/data/models/Event.java (hunk 6 above)
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/app/data/models/Event.java]: Please build the project, it is autogenerated by DBFlow

File: app/src/main/java/org/fossasia/openevent/app/data/models/SimpleModel.java (hunk 8 above)
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/app/data/models/SimpleModel.java]: It is created to test that DBFlow configuration is working, not to be used by outer package

File: app/src/main/java/org/fossasia/openevent/app/common/di/module/DataModule.java (hunk 2 above)
Comment (mejariamol) [inline app/src/main/java/org/fossasia/openevent/app/common/di/module/DataModule.java]: So it will be handled in separate PR right? attendees database

Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/app/common/di/module/DataModule.java]: Yes

File: app/src/main/java/org/fossasia/openevent/app/common/di/module/DatabaseModule.java [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+import dagger.Provides;
+
+@Module
+public class DatabaseModule {
+
+    @Provides
+    IDatabaseRepository providesDatabaseRepository() {
+        return new DatabaseRepository();
+    }
+
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/app/common/di/module/DatabaseModule.java]: I see it is there in all the classes

**Comment by iamareebjamal** [comment]
@arp95 @mejariamol @rohanagarwal94 Please review

File: app/src/main/java/org/fossasia/openevent/app/data/EventRepository.java [hunk 11]
```diff
@@ -16,22 +20,24 @@
...
+import timber.log.Timber;
 
 public class EventRepository implements IEventRepository {
 
     public static final String ORGANIZER = "user";
-    public static final String EVENT = "event";
-    public static final String EVENTS = "events";
     public static final String ATTENDEES = "attendees";
 
+    private IDatabaseRepository databaseRepository;
```
Comment (mejariamol) [inline app/src/main/java/org/fossasia/openevent/app/data/EventRepository.java]: Can't we just use `inject` annotation instead of passing them directly to constructor?

File: app/src/main/java/org/fossasia/openevent/app/data/models/CallForPapers.java [hunk 12]
```diff
@@ -1,20 +1,34 @@
...
     private String startDate;
-    @SerializedName("timezone")
     private String timezone;
 
+    public long getId() {
+        return id;
+    }
+
+    public void setId(long id) {
+        this.id = id;
```
Comment (arp95) [inline app/src/main/java/org/fossasia/openevent/app/data/models/CallForPapers.java]: is long needed for id? i think int values are enough.

File: app/src/main/java/org/fossasia/openevent/app/data/models/Copyright.java [hunk 13]
```diff
@@ -57,12 +71,24 @@ public void setLogo(String logo) {
...
     }
 
-    public Long getYear() {
+    public long getYear() {
         return year;
     }
 
-    public void setYear(Long year) {
+    public void setYear(long year) {
         this.year = year;
```
Comment (arp95) [inline app/src/main/java/org/fossasia/openevent/app/data/models/Copyright.java]: same here.

File: app/src/main/java/org/fossasia/openevent/app/data/models/SimpleModel.java [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+
+    public SimpleModel(long id, String name, String description) {
+        this.id = id;
```
Comment (arp95) [inline app/src/main/java/org/fossasia/openevent/app/data/models/SimpleModel.java]: what is the simple model for?
also are the current fields fine. dont we need to add more?

File: app/src/main/java/org/fossasia/openevent/app/common/di/module/DatabaseModule.java [hunk 15]
```diff
@@ -0,0 +1,17 @@
...
+import org.fossasia.openevent.app.data.db.contract.IDatabaseRepository;
+
+import dagger.Module;
+import dagger.Provides;
+
+@Module
+public class DatabaseModule {
+
+    @Provides
+    IDatabaseRepository providesDatabaseRepository() {
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/app/common/di/module/DatabaseModule.java]: True. Thanks

File: app/src/main/java/org/fossasia/openevent/app/data/models/CallForPapers.java (hunk 12 above)
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/app/data/models/CallForPapers.java]: This is by default from server

File: app/src/main/java/org/fossasia/openevent/app/data/models/SimpleModel.java (hunk 14 above)
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/app/data/models/SimpleModel.java]: This is to test that DBFlow configuration works via DatabaseRepository

Take a look at DatabaseRepositoryTest class

File: app/src/main/java/org/fossasia/openevent/app/data/db/configuration/DbFlowExclusionStrategy.java (hunk 5 above)
Comment (rohanagarwal94) [inline app/src/main/java/org/fossasia/openevent/app/data/db/configuration/DbFlowExclusionStrategy.java]: @iamareebjamal is this done?

Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/app/data/db/configuration/DbFlowExclusionStrategy.java]: Yes
